File: src/com/suushiemaniac/cubing/alglib/move/modifier/ClockModifier.java
package com.suushiemaniac.cubing.alglib.move.modifier;

public class ClockModifier implements Modifier {
    public static final ClockModifier endPinModifier = new ClockModifier(ClockNumModifier.ZERO, ClockDirectionModifier.NEG);

    public static ClockModifier fromNumber(int number) {
        ClockNumModifier numModifier = ClockNumModifier.fromNotation(Math.abs(number) + "");
        ClockDirectionModifier directionModifier = number < 0 ? ClockDirectionModifier.NEG : ClockDirectionModifier.POS;
        return new ClockModifier(numModifier, directionModifier);
    }

    private ClockNumModifier numModifier;
    private ClockDirectionModifier directionModifier;

    public ClockModifier(ClockNumModifier numModifier, ClockDirectionModifier directionModifier) {
        this.numModifier = numModifier;
        this.directionModifier = directionModifier;
    }

    public ClockNumModifier getNumModifier() {
        return this.numModifier;
    }

    public ClockDirectionModifier getDirectionModifier() {
        return this.directionModifier;
    }

    @Override
    public ClockModifier inverse() {
        return new ClockModifier(this.numModifier, this.directionModifier.inverse());
    }

    @Override
    public ClockModifier merge(Modifier other) {
        if (!(other instanceof ClockModifier)) return null;
        int newRot = this.toNumber() + ((ClockModifier) other).toNumber();
        if (newRot < -6) newRot = 6 + (newRot + 6);
        else if (newRot > 6) newRot = -6 + (newRot - 6);
        return ClockModifier.fromNumber(newRot);
    }

    @Override
    public String toFormatString() {
        return this.numModifier.toFormatString() + this.directionModifier.toFormatString();
    }

    public String toNumString() {
        return this.directionModifier.toFormatString() + this.numModifier.toFormatString();
    }

    public int toNumber() {
        return Integer.parseInt(this.toNumString());
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof ClockModifier
                && ((ClockModifier) obj).numModifier == this.numModifier
                && ((ClockModifier) obj).directionModifier == this.directionModifier;
    }

    @Override
    public int hashCode() {
        return this.toFormatString().hashCode();
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/alglib/move/modifier/ClockModifier.kt
package com.suushiemaniac.cubing.alglib.move.modifier

import kotlin.math.abs

data class ClockModifier(val numModifier: ClockNumModifier, val directionModifier: ClockDirectionModifier) : Modifier {
    override val notation = "${this.numModifier.notation}${this.directionModifier.notation}"

    override fun inverse() = ClockModifier(this.numModifier, this.directionModifier.inverse())

    override fun merge(other: Modifier): ClockModifier {
        if (other !is ClockModifier) return other as ClockModifier

        val newRot = (this.toNumber() + other.toNumber()).let {
            when {
                it < -6 -> 6 + (it + 6)
                it > 6 -> -6 + (it - 6)
                else -> it
            }
        }

        return ClockModifier.fromNumber(newRot)
    }

    fun toNumber() = "${this.directionModifier.notation}${this.numModifier.notation}".toInt()

    override fun toString() = this.notation

    companion object {
        val endPinModifier = ClockModifier(ClockNumModifier.ZERO, ClockDirectionModifier.NEG)

        fun fromNumber(number: Int): ClockModifier {
            val directionModifier = if (number < 0) ClockDirectionModifier.NEG else ClockDirectionModifier.POS
            val numModifier = ClockNumModifier.values().find { it.notation == abs(number).toString() }!!

            return ClockModifier(numModifier, directionModifier)
        }
    }
}